jsonrpc/specification.md

```diff
@@ -0,0 +1,49 @@
+# Portal JSON-RPC specification
+
+The `*` is meant to be a placeholder to substitute different Portal sub-networks `history`, `state`, `beacon`, etc
+
+## portal_*RoutingTableInfo
+
+## portal_*AddEnr
+- Adds the Enr to the respective sub-networks routing table
+
+## portal_*GetEnr
+
+## portal_*DeleteEnr
+
+## portal_*LookupEnr
+
+## portal_*Ping
+
+## portal_*FindNodes
+
+## portal_*FindContent
+- must validate content
+
+## portal_*Offer
+- must not validate content
+- shouldn't store data on offering node
+
+## portal_*RecursiveFindNodes
+
+## portal_*GetContent
+- must validate content
+- must check local storage first before doing recursive find content
+- if recursive find content is done must do poke mechanism
```

Comment:
Poke is currently not part of the state network. Not sure if this should be mentioned.

Reply:
> Poke is currently not part of the state network. Not sure if this should be mentioned.

good point I will remove it :rocket:

Reply:
I was under the impression that we could indeed poke the content, since we are effectively building up a proof by walking the trie...  Couldn't the node that is doing the trie walk push the content into the nodes that *should* have had it but didn't after retrieval?

Reply:
> I was under the impression that we could indeed poke the content, since we are effectively building up a proof by walking the trie... Couldn't the node that is doing the trie walk push the content into the nodes that _should_ have had it but didn't after retrieval?

You are right, so I will keep poke listed here

Reply:
Actually poke isn't possible in portal_stateGetContent because we don't have the proof needed in order to build an offer in this context. The higher level rpc endpoints such as eth_getBalance will have this proof from walking the trie but portal_stateGetContent doesn't have a proof parameter and may get called/used in other contexts where we are not walking the trie. Sure, you could in theory walk the trie for every bit a state content being looked up but that would be very slow due to the many network requests potentially required.

Reply:
We could add optional `proof` field to the `portal_stateGetContent`. If provided, it can be used together with content obtained from the network to generate content that can be poked.

This would technically make `get_content` for State different then the one from History and Beacon. But it actually is not that different, because for Beacon and History, you don't need any additional data in order to create poke content item.

I didn't think extensively about it, but it feels like we would have similar need for Transaction and Verkle networks (when added). So State wouldn't be the only network that will need this.

Side note: I'm on the fence if the field should be optional or not. I'm not sure in what situation one would have content key, but not have a proof for it. But on the other hand, I don't want to make API too strict.

Reply:
I don't think we should make API decisions with debugging/testing use case in mind. If we need something specific for debugging, those can be added separately (e.g. `portalDebug_stateGetContent`) and they can be:
1. documented to behave the same as the original request but skipping validation
2. enabled/disabled with a flag

But, if we have legitimate use case when we are requesting data but don't already have proof available, then field should of course be optional. I'm just not sure if such use case exists

Reply:
Since we *sometimes* can poke the content, we should just specify this as a **SHOULD** with that context so that in the event that the call if poke-able it should be poked, otherwise, if it's unanchored content, it shouldn't.